src/tools/kata-ctl/src/arch/x86_64/mod.rs

```diff
@@ -61,4 +68,88 @@ mod arch_specific {
+        let cpu_info = check::get_single_cpu_info(check::PROC_CPUINFO, CPUINFO_DELIMITER)?;
+
+        let cpu_flags = check::get_cpu_flags(&cpu_info, CPUINFO_FLAGS_TAG).map_err(|e| {
+            anyhow!(
+                "Error parsing CPU flags, file {:?}, {:?}",
+                check::PROC_CPUINFO,
+                e
+            )
+        })?;
+        Ok(cpu_flags)
+    }
+
+    pub const TDX_SYS_FIRMWARE_DIR: &str = "/sys/firmware/tdx_seam/";
+    pub const TDX_CPU_FLAG: &str = "tdx";
+    pub const SEV_KVM_PARAMETER_PATH: &str= "/sys/module/kvm_amd/parameters/sev";
+    pub const SNP_KVM_PARAMETER_PATH: &str = "/sys/module/kvm_amd/parameters/sev_snp";
+
+    pub fn available_guest_protection() -> Result<GuestProtection>{
+        return arch_guest_protection(TDX_SYS_FIRMWARE_DIR, 
+            TDX_CPU_FLAG, SEV_KVM_PARAMETER_PATH, SNP_KVM_PARAMETER_PATH);
+    }
+
+    pub fn arch_guest_protection(tdx_path: &str, tdx_flag: &str, sev_path: &str, snp_path: &str) -> Result<GuestProtection>{
+        if !Uid::effective().is_root() {
+            return Err(ProtectionError::GuestProtectNoPerms)?;
+        }
+
+        let flags =  retrieve_cpu_flags()
+            .map_err(|err| ProtectionError::GuestProtectCheckFailed(String::from(err.to_string())))?;
+
+        let metadata = fs::metadata(tdx_path);
+        
+        //TODO: Use code below once `is_ok_and` is stable
+        //if metadata.is_ok_and(|metadata| metadata.is_dir()) 
+
+        if metadata.is_ok() && metadata.unwrap().is_dir() {
+            if flags.contains(tdx_flag) {
+                return Ok(GuestProtection::TDX);
+            }
+        }
+
+        let check_contents = |file_name: &str| -> Result<bool> {
+            let contents = fs::read_to_string(file_name)
+                .map_err(|err| ProtectionError::GuestProtectCheckFailed(format!("Error reading file {} : {}", file_name, err)))?;
+
+            //if contents.is_ok() && contents.unwrap() == "Y" {
+            //    return true;
+            //}
+            if contents == "Y" {
+                return Ok(true);
+            }
+            Ok(false)
+        };
+
+        if check_contents(snp_path)? {
+            return Ok(GuestProtection::SNP);
+        }
+
+        if check_contents(sev_path)? {
+            return Ok(GuestProtection::SEV);
+        }
+
+        return Ok(GuestProtection::NoProtection);
+    }
 }
+
+    mod tests {
+        use super::*;
+        use crate::check;
+    
+        #[test]
+        fn test_available_guest_protection() {
+            let res = available_guest_protection();
+            assert_eq!(res.unwrap(), check::GuestProtection::NoProtection);
+        }
+
+        fn test_arch_guest_protection() {
+            let res = arch_guest_protection(TDX_SYS_FIRMWARE_DIR, 
+                TDX_CPU_FLAG, SEV_KVM_PARAMETER_PATH, SNP_KVM_PARAMETER_PATH);
+            assert_eq!(res.unwrap(), check::GuestProtection::NoProtection);
+        }
```

Comment:
Won't these tests fail on systems with guest protection? Also, I'd add checks to see if the tests are running as root as that will change the expected results in some cases.